processor.go

```diff
@@ -143,7 +130,217 @@ func (p *ServiceProcessor) RegisterStreamingHandler(f interface{}) error {
+// The response is also JSON encoded.
+//
+// For GET requests, the callback is registered on the same URL. But clients
+// can also query individual resources such as
+// /v$version/$namespace/$msgStructName/$id. For this to work, msg in the
+// callback must be a singleton struct with either an integer or a byte slice.
+// For integers, the client can directly query the integer resource, for byte
+// slices, the clients must query the hex encoded representation. Using an
+// empty struct for msg is also supported.
+//
+// The min/maxVersion argument represents the range of versions where the API
+// is present. If breaking changes must be made then they must use a new
+// version.
+//
+// This method is experimental.
+func (p *ServiceProcessor) RegisterRESTHandler(f interface{}, namespace, method string, minVersion, maxVersion int) error {
+	// TODO support more methods
+	if method != "GET" && method != "POST" && method != "PUT" {
+		return errors.New("invalid REST method")
+	}
+	if minVersion > maxVersion {
+		return errors.New("min version is greater than max version")
+	}
+	if minVersion < 3 {
+		return errors.New("earliest supported API level must be greater or equal to 3")
+	}
+	if err := handlerInputCheck(f); err != nil {
+		return err
+	}
+	resource, sh, err := createServiceHandler(f)
+	if err != nil {
+		return err
+	}
+	var k kindGET
+	var fieldNameGET string
+	if method == "GET" {
+		k, fieldNameGET, err = prepareHandlerGET(f)
+		if err != nil {
+			return err
+		}
+	}
+	h := func(w http.ResponseWriter, r *http.Request) {
+		if r.Method != method {
+			http.Error(w, "unsupported method: "+r.Method, http.StatusMethodNotAllowed)
+			return
+		}
+		var msgBuf []byte
+		switch r.Method {
+		case "GET":
+			switch k {
+			case emptyGET:
+				msgBuf = []byte("{}")
+			case intGET:
+				intRegex := fmt.Sprintf(`^/v\d/%s/%s/\d+$`, namespace, resource)
+				ok, err := regexp.MatchString(intRegex, r.URL.EscapedPath())
+				if err != nil {
+					http.Error(w, err.Error(), http.StatusInternalServerError)
+					return
+				}
+				if !ok {
+					http.Error(w, "invalid path", http.StatusNotFound)
+					return
+				}
+				_, num := path.Split(r.URL.EscapedPath())
+				msgBuf = []byte(fmt.Sprintf("{\"%s\": %s}", fieldNameGET, num))
+			case sliceGET:
+				sliceRegex := fmt.Sprintf(`^/v\d/%s/%s/[0-9a-f]+$`, namespace, resource)
+				ok, err := regexp.MatchString(sliceRegex, r.URL.EscapedPath())
+				if err != nil {
+					http.Error(w, err.Error(), http.StatusInternalServerError)
+					return
+				}
+				if !ok {
+					http.Error(w, "invalid path", http.StatusNotFound)
+					return
+				}
+				_, hexStr := path.Split(r.URL.EscapedPath())
+				byteBuf, err := hex.DecodeString(hexStr)
+				if err != nil {
+					http.Error(w, err.Error(), http.StatusBadRequest)
+					return
+				}
+				encoded := base64.StdEncoding.EncodeToString(byteBuf)
+				msgBuf = []byte(fmt.Sprintf("{\"%s\": \"%s\"}", fieldNameGET, encoded))
```

Comment:
I gotta say, I'm really suspicious of this "parse out the GET argument and then reformat it into JSON" technique. It seems to obscure what's really happening here which is "move the data from the GET URL to the request struct". There's already two cases, one for int and one for hex string. So in those two cases do the case-specific reflect garbage, so that once you leave the switch on method, you're ready to call callInterfaceFunc on msg directly (json.Unmarshal would only be called in POST and PUT.